native/src/main/java/io/ballerina/stdlib/crypto/PasswordUtils.java

```diff
@@ -0,0 +1,175 @@
+            );
+        }
+        return null;
+    }
+
+    /**
+     * Generate a cryptographically secure random salt.
+     *
+     * @return byte array containing the generated salt
+     */
+    public static byte[] generateRandomSalt() {
+        byte[] salt = new byte[SALT_LENGTH];
+        SECURE_RANDOM.nextBytes(salt);
+        return salt;
+    }
+
+    /**
+     * Format a BCrypt hash string according to the standard format.
+     *
+     * @param workFactor the work factor used in hashing
+     * @param saltAndHash combined salt and hash byte array
+     * @return formatted BCrypt hash string
+     */
+    public static String formatBCryptHash(long workFactor, byte[] saltAndHash) {
+        String saltAndHashBase64 = Base64.toBase64String(saltAndHash);
+        return String.format("$2a$%02d$%s", workFactor, saltAndHashBase64);
+    }
+
+    /**
+     * Format an Argon2 hash string according to the standard format.
+     *
+     * @param memory memory cost parameter
+     * @param iterations time cost parameter
+     * @param parallelism parallelism parameter
+     * @param saltBase64 Base64 encoded salt
+     * @param hashBase64 Base64 encoded hash
+     * @return formatted Argon2 hash string
+     */
+    public static String formatArgon2Hash(long memory, long iterations, long parallelism, 
+            String saltBase64, String hashBase64) {
+        return String.format("$argon2id$v=19$m=%d,t=%d,p=%d$%s$%s",
+                memory, iterations, parallelism, saltBase64, hashBase64);
+    }
+
+    /**
+     * Format an Argon2 salt string according to the standard format.
+     *
+     * @param memory memory cost parameter
+     * @param iterations time cost parameter
+     * @param parallelism parallelism parameter
+     * @param saltBase64 Base64 encoded salt
+     * @return formatted Argon2 salt string
+     */
+    public static String formatArgon2Salt(long memory, long iterations, long parallelism, String saltBase64) {
+        return String.format("$argon2id$v=19$m=%d,t=%d,p=%d$%s",
+                memory, iterations, parallelism, saltBase64);
+    }
+
+    /**
+     * Combine salt and hash byte arrays into a single array.
+     *
+     * @param salt salt byte array
+     * @param hash hash byte array
+     * @return combined byte array containing salt followed by hash
+     */
+    public static byte[] combine(byte[] salt, byte[] hash) {
+        byte[] combined = new byte[salt.length + hash.length];
+        System.arraycopy(salt, 0, combined, 0, salt.length);
+        System.arraycopy(hash, 0, combined, salt.length, hash.length);
+        return combined;
+    }
+
+    /**
+     * Compare two byte arrays in constant time to prevent timing attacks.
+     *
+     * @param a first byte array
+     * @param b second byte array
+     * @return true if arrays are equal, false otherwise
+     */
+    public static boolean constantTimeArrayEquals(byte[] a, byte[] b) {
```

Comment:
There is also [`MessageDigest#isEqual`](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/security/MessageDigest.html#isEqual(byte[],byte[])) to which this could maybe delegate instead, though its implementation is quite similar.